src/pages/expansions/StocktakeEditExpansion.js

```diff
@@ -0,0 +1,207 @@
+    this.state = {};
+    autobind(this);
+  }
+
+  onEndEditing(key, stocktakeBatch, newValue) {
+    this.props.database.write(() => {
+      switch (key) {
+        case 'countedTotalQuantity': {
+          const newCountedQuantity = parsePositiveInteger(newValue);
+          if (newCountedQuantity === null) return;
+
+          stocktakeBatch.countedTotalQuantity = parsePositiveInteger(newValue);
+          // Should refresh StocktakeEditPage, because no changes are made to
+          // StocktakeItem object, but countedTotalQuantity getter result will be different
+          this.props.refreshParent();
+          break;
+        }
+        case 'batch': {
+          if (!newValue || newValue === '') return;
+          stocktakeBatch.batch = newValue;
+          break;
+        }
+        case 'expiryDate':
+          stocktakeBatch.expiryDate = newValue;
+          break;
+        default:
+          break;
+      }
+      this.props.database.save('stocktakeBatch', stocktakeBatch);
+    });
+  }
+
+  refreshData() {
+    this.setState({ data: this.item.batches });
+  }
+
+  renderCell(key, stocktakeBatch) {
+    const isEditable = !this.item.stocktake.isFinalised;
+    switch (key) {
+      default:
+        return {
+          cellContents: stocktakeBatch[key],
+        };
+      case 'batch':
+        return {
+          type: isEditable ? 'editable' : 'text',
+          cellContents: stocktakeBatch[key] && stocktakeBatch[key] !== '' ?
+                        stocktakeBatch[key] : '(no batch name)', // TODO: localise
+        };
+      case 'countedTotalQuantity': {
+        const emptyCellContents = isEditable ? '' : tableStrings.no_change;
+        return {
+          type: isEditable ? 'editable' : 'text',
+          cellContents: stocktakeBatch.difference === 0 ? emptyCellContents :
+                        stocktakeBatch.countedTotalQuantity,
+          placeholder: tableStrings.no_change,
+        };
+      }
+      case 'expiryDate': {
+        return (
+          <ExpiryTextInput
+            key={key}
+            isEditable={isEditable}
+            onEndEditing={(newValue) => {
+              this.onEndEditing(key, stocktakeBatch, newValue);
+              this.refreshData();
+            }}
+            text={stocktakeBatch[key]}
+            style={dataTableStyles.text}
+          />
+        );
+      }
+    }
+  }
+
+  renderAddBatchButton() {
+    const addNewBatch = () => {
+      this.props.database.write(() => {
+        this.item.createNewBatch(this.props.database);
+      });
+      this.refreshData();
+    };
+    return (
+      <PageButton
+        text={buttonStrings.add_batch}
+        onPress={addNewBatch}
+        isDisabled={this.item.stocktake.isFinalised}
+        style={localStyles.addBatchButton}
+      />
+    );
+  }
+
+  renderFooter() {
+    return null;
+  }
+
+  renderPageInfo() {
+    const infoColumns = [
+      [
+        {
+          title: 'By Batch:',
+          info: this.item.item.name,
```

Comment:
I think `this.item` would be better named as `this.stocktakeItem` in order to avoid this quirk.

Reply:
Is there a limitation stopping props.data just being props.stocktakeItem and using that directly? Or is a benefit to binding it to `this` in the constructor? I'd think that either case is pointing at a object in a live Realm.Results object right?

Reply:
no benefit you are right.